### PR TITLE
Implement shares sale flow

### PR DESCRIPTION
**Pull request by Artem-Semenov-dev**
This PR implements shares sale flow.
The process modeling of the shares sale flow:
image

### CODE REVIEW

File: server/src/main/java/io/spine/examples/shareaware/server/market/MarketProcess.java [hunk 1]
```diff
@@ -29,10 +29,19 @@
...
+import io.spine.examples.shareaware.market.command.SellSharesOnMarket;
 import io.spine.examples.shareaware.market.event.SharesObtained;
+import io.spine.examples.shareaware.market.event.SharesSoldOnMarket;
+import io.spine.money.Currency;
+import io.spine.money.Money;
 import io.spine.server.command.Assign;
 import io.spine.server.procman.ProcessManager;
 
+import java.security.SecureRandom;
+import java.util.Random;
```
Comment (Vladyslav-Kuksiuk) [inline server/src/main/java/io/spine/examples/shareaware/server/market/MarketProcess.java:41]: Please remove redundant imports.

File: server/src/test/java/io/spine/examples/shareaware/server/investment/given/InvestmentTestContext.java [hunk 2]
```diff
@@ -39,14 +40,14 @@
 import io.spine.server.BoundedContextBuilder;
 import io.spine.server.DefaultRepository;
 
-public class PurchaseTestContext {
+public class InvestmentTestContext {
```
Comment (Vladyslav-Kuksiuk) [inline server/src/test/java/io/spine/examples/shareaware/server/investment/given/InvestmentTestContext.java:43]: Looks `final` to me.

File: server/src/test/java/io/spine/examples/shareaware/server/investment/given/SharesSaleTestEnv.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+import io.spine.examples.shareaware.wallet.Wallet;
+import io.spine.examples.shareaware.wallet.command.RechargeBalance;
+import io.spine.examples.shareaware.wallet.event.BalanceRecharged;
+import io.spine.money.Money;
+
+import static io.spine.examples.shareaware.server.wallet.MoneyCalculator.multiply;
+import static io.spine.examples.shareaware.server.wallet.MoneyCalculator.subtract;
+import static io.spine.examples.shareaware.server.wallet.MoneyCalculator.sum;
+
+public class SharesSaleTestEnv {
```
Comment (Vladyslav-Kuksiuk) [inline server/src/test/java/io/spine/examples/shareaware/server/investment/given/SharesSaleTestEnv.java:52]: Looks `final` to me.

**Review by Vladyslav-Kuksiuk — COMMENTED** [review]
@Artem-Semenov-dev almost LGTM, please see my comments.

File: server/src/main/java/io/spine/examples/shareaware/server/investment/InvestmentAggregate.java [hunk 4]
```diff
@@ -57,4 +64,68 @@ private void event(SharesAdded e) {
...
+                .newBuilder()
+                .setInvestment(c.getInvestment())
+                .setProcess(c.getProcess())
+                .setSharesAvailable(state().getSharesAvailable())
+                .vBuild();
+    }
+
+    @Apply
+    private void event(SharesReservationCompleted e) {
+        builder().removeSharesReserved(e.getProcess()
```
Comment (armiol) [inline server/src/main/java/io/spine/examples/shareaware/server/investment/InvestmentAggregate.java:108]: When emitted, `SharesReservationCompleted` includes `shares_available` field. I am not sure why we don't use it in the applier. Or, if it is not needed in this applier, why we have this field in this event at all.

Something does not quite add up here to me.

File: server/src/test/java/io/spine/examples/shareaware/server/investment/SharesSaleTest.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            context().assertEvent(expected);
+        }
+    }
+
+    @Nested
+    @DisplayName("be led by shares sale process")
+    class SaleProcess {
+
+        @Test
+        @DisplayName("with state")
```
Comment (armiol) [inline server/src/test/java/io/spine/examples/shareaware/server/investment/SharesSaleTest.java:190]: Not sure I understand this scenario: 
```
`SharesSale` should ... be led by shares sale process ... with state
```
What does this even mean?

Please review.

File: server/src/test/java/io/spine/examples/shareaware/server/investment/SharesSaleTest.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            SellShares command = sellShareFrom(investment);
+            context().receivesCommand(command);
+            SharesSold expected = sharesSoldBy(command, investment);
+
+            context().assertEvent(expected);
+        }
+
+        @Test
+        @DisplayName("which emits the `SharesSaleFailed` " +
+                "due to insufficient shares in the investment")
```
Comment (armiol) [inline server/src/test/java/io/spine/examples/shareaware/server/investment/SharesSaleTest.java:270]: "...insufficient quantity of shares owned"?

File: model/src/main/proto/spine_examples/shareaware/investment/sale_commands.proto [hunk 7]
```diff
@@ -51,6 +51,9 @@ message SellShares {
     // The ID of the share to sell.
     ShareId share = 3 [(required) = true];
 
+    // Wanted price per one share.
```
Comment (armiol) [inline model/src/main/proto/spine_examples/shareaware/investment/sale_commands.proto:54]: What we lack here is the description of the case, in which (theoretically) this "wanted" price does not correlate with the actual price on market.

I mean, the problem here is with this "wanted" word. If we say "wanted" we mean "this is the price I want my shares to be sold for", which implies the case in which "actual" price may differ (and hence, some rejections from `Market` may be thrown).

I suggest to describe this case, or find another word for "wanted".

**Review by armiol — CHANGES_REQUESTED** [review]
@Artem-Semenov-dev please see my comments.

File: server/src/main/java/io/spine/examples/shareaware/server/investment/InvestmentAggregate.java (hunk 4 above)
Comment (Artem-Semenov-dev) [reply server/src/main/java/io/spine/examples/shareaware/server/investment/InvestmentAggregate.java:108]: But the information about available shares is needed for the `HeldShares` projection and I cannot get this information from anywhere else. Of course, I can calculate the amount of the currently available shares in the projection, but you told me not to do it in some previous PR.

**Review by armiol — APPROVED** [review]

**Pull request merged**